### PR TITLE
Zbmath fetcher

### PR DESCRIPTION
**Pull request by ibe-314**
Add zbmath to the public databases which can be chosen when updating an entry with bibliographic information from the web. See issue [#7437].

Screenshot (3)

Screenshot (4)

- [x] Change in CHANGELOG.md described (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by ibe-314** [comment]
In the Fetcher Tests  src/test/java/org/jabref/logic/importer/fetcher/ZbMATHTest.java did not fail but some tests for other fetchers. I think this failures are not due to my changes.

File: src/main/java/org/jabref/logic/importer/EntryBasedParserFetcher.java [hunk 1]
```diff
@@ -54,6 +55,16 @@ default void doPostCleanup(BibEntry entry) {
     default List<BibEntry> performSearch(BibEntry entry) throws FetcherException {
         Objects.requireNonNull(entry);
 
+        try {
+            if (getURLForEntry(entry) == null) {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/importer/EntryBasedParserFetcher.java:59]: This now calls `getURLForEntry` twice, better extract it to a variable.

File: src/main/java/org/jabref/logic/importer/EntryBasedParserFetcher.java [hunk 2]
```diff
@@ -54,6 +55,16 @@ default void doPostCleanup(BibEntry entry) {
...
+                return Collections.emptyList();
+            }
+        } catch (MalformedURLException e) {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/importer/EntryBasedParserFetcher.java:62]: You can combine these catch blocks by using `catch(MalformedURLException|URISyntaxException e)`

File: src/main/java/org/jabref/logic/importer/fetcher/ZbMATH.java [hunk 3]
```diff
@@ -40,17 +46,55 @@ public String getName() {
...
+        entry.getFieldOrAlias(StandardField.JOURNAL).ifPresent(journal -> uriBuilder.addParameter("j", journal));
+        entry.getFieldOrAlias(StandardField.YEAR).ifPresent(year -> uriBuilder.addParameter("y", year));
+        entry.getFieldOrAlias(StandardField.PAGINATION)
+             .ifPresent(pagination -> uriBuilder.addParameter("p", pagination));
+        entry.getFieldOrAlias(StandardField.VOLUME).ifPresent(volume -> uriBuilder.addParameter("v", volume));
+        entry.getFieldOrAlias(StandardField.ISSUE).ifPresent(issue -> uriBuilder.addParameter("i", issue));
+
+        if (entry.getFieldOrAlias(StandardField.AUTHOR).isPresent()) {
+            // replace "and" by ";" as citation matching API uses ";" for separation
+            String author = entry.getFieldOrAlias(StandardField.AUTHOR).get();
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/importer/fetcher/ZbMATH.java:71]: Please use the `AuthorList` class for parsing and reformatting the authors. This also handles a few edge cases such as `author={{United and Free States}}` which is one author (an "organization") and not two.

File: src/main/java/org/jabref/logic/importer/fetcher/ZbMATH.java [hunk 4]
```diff
@@ -40,17 +46,55 @@ public String getName() {
...
+            author = author.replace(" and ", ";");
+            uriBuilder.addParameter("a", author);
+        }
+
+        String urlString = uriBuilder.build().toString();
+        HttpResponse<JsonNode> response = Unirest.get(urlString)
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/importer/fetcher/ZbMATH.java:77]: Can you please add a comment that the API does not provide a means to return bibtex directly, and we thus extract the zbl id and use getUrlForIdentifier which does return bibtex. 

Btw, could you please contact zbl and maybe ask if they could add a `format=bibtex` query parameter to their match api? That would make this additional roundtrip unnecessary. Thanks!

File: src/test/java/org/jabref/logic/importer/fetcher/ZbMATHTest.java [hunk 5]
```diff
@@ -44,18 +43,42 @@ void setUp() throws Exception {
...
         donaldsonEntry.setField(StandardField.YEAR, "1983");
-        donaldsonEntry.setField(new UnknownField("zbl"), "0507.57010");
+        donaldsonEntry.setField(StandardField.ZBL_NUMBER, "0507.57010");
     }
 
     @Test
     void searchByQueryFindsEntry() throws Exception {
         List<BibEntry> fetchedEntries = fetcher.performSearch("an:0507.57010");
-        assertEquals(Collections.singletonList(donaldsonEntry), fetchedEntries);
+        assertEquals(donaldsonEntry, fetchedEntries.get(0));
```
Comment (tobiasdiez) [inline src/test/java/org/jabref/logic/importer/fetcher/ZbMATHTest.java:52]: Why did you change this? The search should only return one entry, or not?

File: src/test/java/org/jabref/logic/importer/fetcher/ZbMATHTest.java [hunk 6]
```diff
@@ -44,18 +43,42 @@ void setUp() throws Exception {
...
     }
+
+    @Test
+    void searchByEntryFindsEntry() throws Exception {
+        BibEntry searchEntry = new BibEntry();
+        searchEntry.setField(StandardField.TITLE, "An application of gauge theory to four dimensional topology");
+        searchEntry.setField(StandardField.AUTHOR, "S. K. {Donaldson}");
+
+        List<BibEntry> fetchedEntries = fetcher.performSearch(searchEntry);
+        assertEquals(donaldsonEntry, fetchedEntries.get(0));
```
Comment (tobiasdiez) [inline src/test/java/org/jabref/logic/importer/fetcher/ZbMATHTest.java:68]: As above, I would prefer if that would compare against `Collections.singletonList(donaldsonEntry)`

File: src/test/java/org/jabref/logic/importer/fetcher/ZbMATHTest.java [hunk 7]
```diff
@@ -44,18 +43,42 @@ void setUp() throws Exception {
...
+
+        searchEntry.setField(StandardField.TITLE, "t");
```
Comment (tobiasdiez) [inline src/test/java/org/jabref/logic/importer/fetcher/ZbMATHTest.java:70]: Please put this second test into a new test method.

File: src/test/java/org/jabref/logic/importer/fetcher/ZbMATHTest.java [hunk 8]
```diff
@@ -44,18 +43,42 @@ void setUp() throws Exception {
...
+        assertEquals(Collections.emptyList(), fetchedEntries);
+    }
+
+    @Test
+    void searchByIdInEntryFindsEntry() throws Exception {
+        BibEntry searchEntry = new BibEntry();
+        searchEntry.setField(StandardField.ZBL_NUMBER, "0507.57010");
+
+        List<BibEntry> fetchedEntries = fetcher.performSearch(searchEntry);
+        assertEquals(donaldsonEntry, fetchedEntries.get(0));
```
Comment (tobiasdiez) [inline src/test/java/org/jabref/logic/importer/fetcher/ZbMATHTest.java:82]: Again, `Collections.singletonList(donaldsonEntry)`

**Review by tobiasdiez — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution. This is a nice feature!

I have a few remarks, but they are only minor once.

File: src/main/java/org/jabref/logic/importer/fetcher/ZbMATH.java (hunk 3 above)
Comment (ibe-314) [reply src/main/java/org/jabref/logic/importer/fetcher/ZbMATH.java:71]: Do you mean this class [https://github.com/JabRef/jabref/blob/master/src/main/java/org/jabref/model/entry/AuthorList.java](https://github.com/JabRef/jabref/blob/master/src/main/java/org/jabref/model/entry/AuthorList.java)?

I need the authors separated by semicolon (lastname before firstname). Is there a function doing this? I have not found one. If it not exists, I guess I have to add one, similar to getAsLastFirstNamesWithAnd(), replacing " and " in line 480 by "; "?

File: src/main/java/org/jabref/logic/importer/fetcher/ZbMATH.java (hunk 4 above)
Comment (ibe-314) [reply src/main/java/org/jabref/logic/importer/fetcher/ZbMATH.java:77]: We work on a new citation matching algorithm and we will add bibtex as an output format there. I will change the jabref code if this happens. Right now, I guess it is okay to use this workaround.

File: src/test/java/org/jabref/logic/importer/fetcher/ZbMATHTest.java (hunk 5 above)
Comment (ibe-314) [reply src/test/java/org/jabref/logic/importer/fetcher/ZbMATHTest.java:52]: I had some problems with the tests, and changed that because I thought this was the cause but it was not. I will change it back to assertEquals(Collections.singletonList(donaldsonEntry), fetchedEntries) and in the other tests, too.

**Comment by ibe-314** [comment]
I added the requested changes. Three fetcher Tests from other fetchers failed but not from the zbmath fetcher.

**Review by tobiasdiez — APPROVED** [review]
Thanks a lot for the quick follow-up. The code looks good to me.

Just out of interest, are you affiliated in some way with zentralblatt?

**Comment by ibe-314** [comment]
I work for them ;-) I wrote this sometime ago on gitter...but you might not follow all messages there.
